internal/era/era.go

```diff
@@ -0,0 +1,343 @@
+	metadata metadata // start, count info
+}
+
+type ReadAtSeeker interface {
+	io.ReaderAt
+	io.Seeker
+}
+
+// NewReader returns a new Reader instance.
+func NewReader(r ReadAtSeeker) (*Reader, error) {
+	length, err := r.Seek(0, io.SeekEnd)
+	if err != nil {
+		return nil, err
+	}
+	m, err := readMetadata(r, length)
+	if err != nil {
+		return nil, err
+	}
+	return &Reader{
+		r:        r,
+		e:        e2store.NewReader(r),
+		next:     m.start,
+		length:   length,
+		metadata: m,
+	}, nil
+}
+
+// readOffset reads a specific block's offset from the block index. The value n
+// is the absolute block number desired.
+func (r *Reader) readOffset(n uint64) (int64, error) {
+	var (
+		firstIndex  = -8 - int64(r.metadata.count)*8      // size of count - index entries
+		indexOffset = int64(n-r.metadata.start) * 8       // desired index * size of indexes
+		offOffset   = r.length + firstIndex + indexOffset // offset of block offset
+	)
+	r.clearBuffer()
+	if _, err := r.r.ReadAt(r.buf[:], offOffset); err != nil {
+		return 0, err
+	}
+	// Since the block offset is relative from its location + size of index
+	// value (8), we need to add it to it's offset to get the block's
+	// absolute offset.
+	return offOffset + 8 + int64(binary.LittleEndian.Uint64(r.buf[:])), nil
+}
+
+// Read reads one (block, receipts) tuple from an Era1 archive.
+func (r *Reader) Read() (*types.Block, types.Receipts, error) {
+	block, receipts, err := r.ReadBlockAndReceipts(r.next)
+	if err != nil {
+		return nil, nil, err
+	}
+	r.next += 1
+	return block, receipts, nil
+}
+
+func skipN(r *Reader, off int64, n int) (int64, error) {
+	for i := 0; i < n; i++ {
+		length, err := r.e.LengthAt(off)
+		if err != nil {
+			return off, err
+		}
+		off += length
+	}
+	return off, nil
+}
+
+// readBlob reads an entry of data.
+func (r *Reader) readEntry(n uint64, skip int) (*e2store.Entry, error) {
+	if n < r.metadata.start || r.metadata.start+r.metadata.count < n {
+		return nil, fmt.Errorf("request out-of-bounds: want %d, start: %d, count: %d", n, r.metadata.start, r.metadata.count)
+	}
+	// Read the specified block's offset from the block index.
+	off, err := r.readOffset(n)
+	if err != nil {
+		return nil, fmt.Errorf("error reading block offset: %w", err)
+	}
+	// Skip to the requested entry.
+	if off, err = skipN(r, off, skip); err != nil {
+		return nil, err
+	}
+	// Read entry.
+	var entry e2store.Entry
+	if _, err := r.e.ReadAt(&entry, off); err != nil {
+		return nil, err
+	}
+	return &entry, nil
+}
+
+// ReadHeader reads the header number n RLP.
+func (r *Reader) ReadHeaderRLP(n uint64) ([]byte, error) {
+	e, err := r.readEntry(n, 0)
+	if err != nil {
+		return nil, err
+	}
+	if e.Type != TypeCompressedHeader {
+		return nil, fmt.Errorf("expected header entry, got %x", e.Type)
+	}
+	return io.ReadAll(snappy.NewReader(bytes.NewReader(e.Value)))
+}
+
+// ReadBodyRLP reads the block body number n RLP.
+func (r *Reader) ReadBodyRLP(n uint64) ([]byte, error) {
```

Comment:
The `ReadXXRLP` methods, they're not actually required externally..? If so, better make them unexported for now IMO